# Numeric Types

- Integers (Z) 0,1,2,3 => int
- Rational Numbers (Q) -> Fraction of Integers => fractions.Fractions 2/5
- Real Numbers (R) -> Float, decimal.Decimal (pro větší kontrolu u přesnosti)
- Complex NUmbers (C)--> 3+4J -> Complex
- Boolean values - 0(False), 1(True) - tohle je zajímavější :)

## Integers data type

Ukládaná jako base-2 binárně. 

### 8 BITS

![Loss](p1-1.png)

10011 binárně tzn. base 2 = base 10 - 19 (normal co máme)
V podstatě máme 2 na 0tou atd..z toho mám čísla a pokud nastane 0 tak tu hodnotu nepřebírá a pokud 1 tak ano. Lehoučké :)

Potřebovali jsme 5 bitů abychom uložili číslo 19

A zde nám začíná logika 8 bitů, 16 bitů apod :). Takže maximum je 255. Pokud dám všude 1ky a budu mít tedy 8 bitů využitých. 128+64+32+16+8+4+2+1 (11111111)

Tak a pokud se jedná o záporná čísla musím vyhradit jedno místo pro -. A tudíž mám jen 7 míst a maximální číslo které získám je -128

- 16 bit má rozmezí -32,768 až 32,767
- u 32 bitů mám již -2,147,483,648 ..2,147,483,647 :)
- pokud nemám určený typ (+ nebo -) tak máme [0,2 na 32] => [0...4,294,967,296]
- proto 32 bitové systémy máme maximum 4,294,967,296 bytes => 4,096MB - 4GB :))
- jinak 1 bit má 8 bytů

V Javě, C musíme předem definovat jaké integer chceme mít - koélik bitů budou využítvat.

- byte - 8 bit number -128 - 127
-  short 16 bit ...
- int 32 bit
- long - 64 bit atd..ale Python tak nefunguje

Python používá různý počet bitů. Prostě podle potřeby si určí jak chce. Samozřejmě platí jasná logika větší čísla - více paměti - pomalejší výpočty.

In [2]:
print(type(100)) #vytváříme classu int

<class 'int'>


In [3]:
import sys

In [4]:
sys.getsizeof(0) #používá 24 bytů - OVERHEAD objekt sám o sobě něco vyžaduje

24

In [5]:
sys.getsizeof(1) 

28

In [6]:
sys.getsizeof(2**1000) #160 bytů

160

In [7]:
(160 - 24) * 8 # 1088 bitů pro uložení hodnoty

1088

In [8]:
import time

In [11]:
def calc(a):
    for i in range(10_000_000):
        a * 2

In [12]:
start = time.perf_counter()
calc(10)
end = time.perf_counter()
print(end - start)

0.6151472999990801


In [14]:
start = time.perf_counter()
calc(2**100)
end = time.perf_counter()
print(end - start)

1.1870016000029864


In [15]:
start = time.perf_counter()
calc(2**10000)
end = time.perf_counter()
print(end - start)

5.643072200000461


Menší ukázka, na velikosti výrazně záleží. Dochází samozřejmě ke zpomalení.

## Integers Operations

### Teorie

Hlavní otázka je, jaký je výstup našich operací?

- int + int --> int
- int - int --> int
- int * int --> int
- int ** int --> int
- int / int --> float 3/4 - 0,75 | 10/2 - 5 ALE i tak jako FLOAT
- dělení má vždy výstup float
- 155 / 4 ale jako zlome => numerator / denominator = 38 se zbytkem 3
- a tím máme // dělení => výsledek je číslo beze zbytku
- ale % takto počítáme pouze zbytek
- 155 = 4 * (155 // 4) + (155 % 4) => nebo 4*38+3 :)
- // div - floor division 
- % mod - modulo operator
- n = d * (n // d) + (n % d)

Co je tedy floor divison? 

- největší číslo v levo od našeho čísla
- floor(3,14) -> 3, 1,999 -> 1, floor(2) -> 2
- floor(-3,1) -> -4 
- a // b = floor(a/b)

In [16]:
#Menší počítání pro +
# a = 135
# b = 4
# 135 / 4 == 33,75 (33 a 3/4)
# 135 // 4  -> 33
# 135 % 4 -> 3
#což je samozřejmě
# 33*4 + 3 => 135

In [17]:
#Negativní čísla
# a = -135
#b = 4
# -135 / 4 = -33,75 (-33 a 3/4)
# -135 // 4 -> -34
#No tohle je už trochu problém :D
# -135 % 4 -> 1

#Mod není remainder - jak se všude říká
#ALE musí se rovnat rovnice

# 4 * (-135 // 4) + (-135 % 4)
# 4* - 34 +1
# -135

# V podstatě jde o to že jakmile vezme 33,75 - nejmenší největší je 33
#ale u -33,75 je to -34 a to vše mění

In [ ]:
# Trochu dále
# a = 13, b = 4 --> 13/4 = 3.25 -----> // = 3 ------> % 1
# a = -13 b = 4 --> -13/4 = -3,25 -----> // =  -4------> % 3
# a = 13, b = -4 --> 13/-4 = -3,25 -----> // = -4------> % -3
# a = -13, b = -4 --> -13/-4 = 3,25 -----> // = 3------> % -1

# a = b * (a//b) + a % b

### Code

In [21]:
type(1+1), type(2/1), type(4-10), type(3*2)
#u dělení získáváme float

(int, float, int, int)

In [22]:
print(2/1) 

2.0


In [23]:
import math

In [27]:
math.floor(3.15), math.floor(1.999), math.floor(-3.1)
#tady vidíme náš floor v akci

(3, 1, -4)

In [28]:
math.floor(-3.0000000000001)

-4

In [30]:
math.floor(-3.0000000000000001)
#float mí limitovanou přesnost - takže 1ku už ignoruje

-3

In [36]:
a, b = 33, 16
print(a/b), print(a//b), print(math.floor(a/b))

2.0625
2
2


(None, None, None)

In [37]:
a, b = -33, 16
print(a/b), print(a//b), print(math.floor(a/b))

-2.0625
-3
-3


(None, None, None)

Pouze ukázka naší teorie.

In [ ]:
a = b * (a//b) + (a % b)

In [41]:
#Dělíme / zbytkujeme a dělíme beze zbytku a kontrolujeme rovnici
a, b = 13, 4
print(f"{a} / {b} = {a/b}")
print(f"{a} // {b} = {a//b}")
print(f"{a} % {b} = {a%b}")
print( a == b * (a//b) + (a%b))

13 / 4 = 3.25
13 // 4 = 3
13 % 4 = 1
True


In [42]:
#Dělíme / zbytkujeme a dělíme beze zbytku a kontrolujeme rovnici
a, b = -13, 4
print(f"{a} / {b} = {a/b}")
print(f"{a} // {b} = {a//b}")
print(f"{a} % {b} = {a%b}")
print( a == b * (a//b) + (a%b))

-13 / 4 = -3.25
-13 // 4 = -4
-13 % 4 = 3
True


Znovu potvrzení celé rovnice a teorie.

## Constructors and Bases - Teorie

Integer je třída int. Int classa obsahuje dva parametry. a = int(10) máme 10, pokud zadáme int(10.9) ztratíme 0.9 a máme znovu pouze 10.

- int(True) => 1 a naopak, tzn Boolean jsou též integer
- int(Decimal("10.9) --> a = 10
- Můžeme zadávat i ve stringu a Python převede

Druhý konstruktor Number Base

- int("123") -> Python pozná, že se nám jedná o 123 - BASE10
- BASE parametr je mezi 2 až 36, jinak je defaultně 10
- int("1010", 2) - chceme BASE 2  :) a máme 10 o base 10 (8-0-2-0)
- lepší zápis je int("1010", base = 2) --> udávat base 
- int("A12F", base = 16) -> 41263 v BASE10 - už začíná být zábava
- pozn. base 16 není case sensitive
- máme 10 čísel a 26 znaků --> proto base 36 naše maximum
- jdeme od 0-9 a A - Z tzn. Base 2 = 0,1
- BASE 10 = 0,1,2,3,4,5,6,7,8,9
- BASE 11 => 0-9 a A :) 

Máme bult-in funkce pro získání procesu z BASE10 získat jiný base. 

- bin() - BASE2 binary --> bin(10) --> "0b1010" (0b-pouze pro označení)
- oct() - BASE8 - prefix 0o
- hex() - BASE16 - prefix 0x
- Prefixy jsou zde abychom poznali o jaká base se jedná
- a= 0b1010 --> díky prefixu ví že máme BASE10

Co další BASES? Python sám o sobě ne, už je třeba Custom code pro tento účel.

Zkusíme náš custom code pro převod

In [ ]:
# n = 232, b = 5 chceme zjistit jaké čísla musíme použít pro base5
# doplníme do rovnice
# (232 // 5) * 5 + 232 % 5 = 46 x 5  +2 = 46 * 5 + 2
#Problém je že 46 se nevejde do base..takže rozbereme 46
# 9 - 1 - 2 //9 příliš velké jdeme dále
#  1 4 1 2
# --------> Máme base5 pro 232 je 1412 (neřešit tento postupný koloběh)
#v podstatě všechny čísla dokud nespadali do 5ky - jelo se podle vzroce stále dokola

### Coding

Je nutné vzít v potaz že pro každý base - budeme mít jinou mapu. 12 BASE má mapu 0-9 a A-C apod.

In [45]:
import fractions

In [46]:
a = fractions.Fraction(22,7)

In [49]:
a

Fraction(22, 7)

In [50]:
print(a)

22/7


In [52]:
float(a)

3.142857142857143

In [53]:
int(a)

3

In [54]:
int("101", base = 2)

5

In [56]:
int("FF", base = 16)

255

In [57]:
int("ff", base = 16)

255

Opáčko prozatím.

In [58]:
int("B", base = 11)
#B v base 11 neexistuje

ValueError: invalid literal for int() with base 11: 'B'

In [59]:
bin(10) #číslo 10 v base = 2

'0b1010'

In [60]:
oct(10) #číslo b base base = 8

'0o12'

In [61]:
hex(255) #číslo v base base = 16

'0xff'

In [63]:
b = 0b101 
b

5

Pokud používám prefix tak nemusím používat žádné spešl zápisy.

Nyní na náš custom changer

In [64]:
def from_base10(n, b):
    if b < 2:
        raise ValueError("Base b must be >= 2")
    if n < 0:
        raise ValueError("Number n must b >=0")
    if n == 0:
        return [0]
    digits = []
    while n > 0:
        #m = n % b
        #n = n // n náhražka za divmod
        n, m = divmod(n,b)
        digits.insert(0,m)
    return digits

In [65]:
from_base10(10,2)
#ona se točí dokola protože n se mění
#tzn vezme n a jede dále
#až dokud n nstane 0 tak končí

[1, 0, 1, 0]

In [67]:
from_base10(255,16) 
#všechno funknční 

[15, 15]

Nyní musím encodovat znaky

In [68]:
def encode(digits, digit_map):
    if max(digits) >= len(digit_map):
        raise ValueError("Digits map is not long enough to encode digits")
    encoding = ""
    for d in digits:
        encoding += digit_map[d]
    #projedu digitmap a to co mám v digits a přiřadím podle toho
    return encoding

In [69]:
encode([15,15], "0123456789ABCDEF")

'FF'

In [70]:
def encode(digits, digit_map):
    if max(digits) >= len(digit_map):
        raise ValueError("Digits map is not long enough to encode digits")
    return "".join([digit_map[d] for d in digits])
    #pomocí list comprehension 
    #tzn přidej d z digimapy pokud je v digits 

In [71]:
encode([15,15], "0123456789ABCDEF")

'FF'

Trochu snadnější a též funkční.

In [74]:
def rebase_from10(number, base):
    digit_map = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    if base < 2 or base > 36:
        raise ValueError("Invalid base:2 <=2 base <=36")
    sign = -1 if number < 0 else 1 #hledám záporná čísla
    number *= sign #a změním číslo násobením -1 => na kladné
    
    digits = from_base10(number,base)
    encoding = encode(digits, digit_map)
    if sign == -1:
        encoding = "-"+ encoding
    return encoding
#naše metoda je finální
#sign používám pro změnu na kladné
#dále používám obě funkce výše

In [76]:
e = rebase_from10(10,2)
print(e) #na binary
print(int(e, base = 2)) #a zpátky

1010
10


In [77]:
e = rebase_from10(-10,2)
print(e) #na binary
print(int(e, base = 2)) #a zpátky

-1010
-10


Všechno je funkční a poměrně zřetelné. Menší únava takže bez delších komentářů.

## Rational Numbers

1/2 , -22/7 apod. v podstatě zlomky. Každé číslo, které končí konečným číslem po tečce je považován za racionální číslo. Tzn. například i 0.123456789 je racionální, 8.3/4 - též.

Fraction classa

- from fractions import Fraction
- x = Fraction(3,4) - takto vytváříme frakce - jinak nevznikají
- Frakce se automaticky zredukují pokud jej udáme špatně, tzn například 10/20 si přepíšem jako 1/2
- Numerator musí mít -
- Fraction(numerator=0, demoninator=1) - defaultné hodnmnoty
- Můžeme tvořit frakce z floatu, decimali, stringu, jiné frakce
- Fraction("10) - oprava na Fraction (10,1) default hodnoty
- Fraction("0.125") -> 1/8
- Fracition("22/7") -> Fraction(22,7)

Aritmetické operace

- +,-,*,/
- Pozor na konce a na změnu 
- 2,3 * 1,2 = 1,3 vynásobil
- 2,3 + 1,2 -> 7,6
- Klasická matematika zlomků tedy
- NUMERATOR - DENOMINATOR 

Každý float může být zapsán jako frakce. Fraction(0.75) ->(3,4) ..atd

Fraction se snaží generovat i hodnoty pro iracionální čísla, ale jen odhady. 

1/8 má přesnou reorezentaci floatu - 0.125, ale 3/10 vyhodí divná čísla, pamět si neuloží správně a neudá přesnou hodnotu. Dodá 0.2999999888 a tudíž se nepropojí.

### Code part

In [80]:
from fractions import Fraction

In [81]:
Fraction(1)

Fraction(1, 1)

In [82]:
Fraction(2,1)

Fraction(2, 1)

In [84]:
Fraction(numerator=1, denominator=2)

Fraction(1, 2)

In [85]:
Fraction('0.125')

Fraction(1, 8)

In [87]:
x = Fraction('22/7')
y = Fraction('22/7')

In [88]:
x+y

Fraction(44, 7)

In [89]:
x*y

Fraction(484, 49)

Proste klasika, mame zakladni operace a tvorbu frakci - zlomku.

In [91]:
x.numerator, x.denominator #muzeme zavolat jednotlive casti

(22, 7)

In [92]:
import math
x = Fraction(math.pi)
x

Fraction(884279719003555, 281474976710656)

Tohle uz je pomerne otravna matematicka cast, asi netreba tolik hrotit prozatim.

In [93]:
a = 0.125
print(a)

0.125


In [94]:
b = 0.3
print(b)

0.3


In [95]:
Fraction(a)

Fraction(1, 8)

In [96]:
Fraction(b)

Fraction(5404319552844595, 18014398509481984)

O co zde jde? Neni schopny presne zobrazit cislo 0.3 ale ukazuje jako 0.00000xxx atd..a tudiz nam vyjde psycho zlomek.

In [97]:
format(a, '0.25f') #takze zde je vse ok

'0.1250000000000000000000000'

In [99]:
format(b,'0.25f') #ale zde vidime ze nase cislo neni presne 0.3 a tudiz se zlomek napocita trochu vice 

'0.2999999999999999888977698'

In [100]:
x = Fraction(0.3)
x.limit_denominator(10)

Fraction(3, 10)

In [102]:
float(x)

0.3

Pouze omezuji pocty. Tzn jake nejblizsi pi obdrzim pokud mohu max .10 mist

## Float Internal Representations - teorie

Defaultni hodnota pouzivana Pythonem pro realna cisla. CPython implementuje Float jako C Double - binary 64. Narozdil od integeru, ktery ma variabilni mnozstvi bytu podle potreby. Floats ma fixne 8 bytu a 64 bits + 16 bytes ve forme overhead.

- 1.5E -5 --> 1.5 x 10 na -5
- sign ma 1 bit, exponent 11 bits [-1022,1023] rozmezi pro exponent
- significant digits -> proste pocet cislic nic vic 1.2345, 1234.5 oboji ma 5 sd
- exponent se nepocita jako significant digits

Realna cisla muzeme zobrazit tedy v base 10 jako frakce, decimal apod..jinak neresim

- v binary takto
- 0.11 v base = 2 je 1x2 na -1 a 1x 2na -2 => 0.75 ..2 na -1 = 0.5 atd..
- tzn. zaciname od 0.5 a jdeme dale po mensich
- 0.1101 v base = 2 mame 0.8125 
- rovnice netreba resit 
- mame urcita uskali napr pri snaze zapsat 0.1 binarne se dostaneme pouze na 0.0999755859375
- Tady asi zacinaji problemy s nepresnymi cisly? Zajimave 
- Nejaka cisla maji konecnou hodnotu v base=10 decimal ale v binary nemaji ..
- napriklad 0.75 v base 10 je 0.11 v binary => mame tak konecnou hodnotu a bude cislo presne
- ale napriklad 0.1 v base10 ma nekonecnou hodnotu v binary 2 - takze mame pouze odhadovanou hodnotu

### Code part

In [103]:
# help(float) jasne

In [104]:
float(10)

10.0

In [105]:
float(10.4)

10.4

In [106]:
float('12.5')

12.5

In [109]:
from fractions import Fraction
a = Fraction('22.7')
float(a)

22.7

In [110]:
print(0.1)

0.1


In [112]:
format(0.1,'.25f') #pouze znovu ukazka teorie,ze neketera cisla nejsou presna

'0.1000000000000000055511151'

In [113]:
a = 0.1+0.1+0.1
b = 0.3
a == b

False

Tohle je opravdu shitty ..nemaji presnou hodnotu.

### Floats Equality testing

Vytvari bugy pokud si to neumime hlidat. V nitru jsou ulozeny jako binary a zde nemusi nastat vzdy shoda. 

- round(0.1,1) + round(0.1,1)+ round(0.1,1) == round(0.3,1) --> FALSE
- round(0.1+0.1+0.1,5) == round(0.3,5) -> True
- zaokrouhlovani ale neni reseni, neni to pruzne
- Spravne reseni je dat odchylku tyn..pokud je jejich odchylka napriklad mensi nez 0.00005 tak je budeme povazovat za shodne. 
- Tez lze resit jako procentualni tolerance - vhodnejsi ale ta ma problemy s hodnotz blizko 0
- idelani je pocitat obe hodnoty a pouzit tu vetsi toleranci pro nasi rovnost
- math.isclose je jiz funkcni metoda pro tento ucel je ale nutne udavat absolutni hodnotu v parametru jinak budou problemy u hodnot blize 0
- pro porovnani float cisel je tedy mnohem lepsi pouzivat math.isclose
- a obecne pokud chci porovnanvat hodnoty je lepsi zvolit jiny format

### Code part

In [115]:
x = 0.1+0.1+0.1
y = 0.3
x == y

False

In [116]:
format(x,".25f")

'0.3000000000000000444089210'

In [118]:
format(y,".25f")

'0.2999999999999999888977698'

Nemame shodu, jasne.

In [120]:
round(x,3) == round(y,3)
#jasne zaokrouhlene uz jsou ok ale nevim od jakeho hodnoty jsou shodne

True

In [121]:
x = 10000.01
y = 10000.02
x = 0.01
y - 0.02
#absolutni rozdil stejny ale relativni rozdil vyrazne odlisnz
#proto je nutne brat v potaz absolutni i relativni vztah
#a samozrejme zde je problem se zaokrouhlovanim muze dochazet ke zkresleni

10000.0

In [123]:
from math import isclose #nase trupr metoda pro ladeni float cisel

In [124]:
x = 0.1+0.1+0.1
y = 0.3

In [126]:
isclose(x,y), x==y #jasne metoda je zdatnejsi

(True, False)

In [129]:
x = 123456789.01
y = 123456789.02
isclose(x,y)

True

In [ ]:
x = 0.01
y = 0.01

In [130]:
isclose(x,y)

True

Vidim ze druha moznost rozhodne nema shodu ale isclose i tak ukaze ze jsou shodne.

In [132]:
isclose(x,y, rel_tol=0.01) #musim presneji definovat parametry

True

In [135]:
x = 0.000000001
y = 0.000000002
isclose(x,y, rel_tol=0.01)

False

Mela by byt shoda, u malych cisel je nutna spise absolutni hodnoty jinak nastanou problemy.

In [136]:
x = 0.000000001
y = 0.000000002
isclose(x,y, rel_tol=0.01, abs_tol=0.01)

True

A ted je vse ok - takze kombinovat relativni i absolutni tolerance is way to go.

### Floats coercing to integer - teorie

Pokud provádíme přeměnu z float hodnoty na integer. Máme stím spojené ztráty. Je otázka zda chcem "zaokrouhlit" hodnotu nahoru nebo dolu. Máme tedy truncation, floor, ceiling a rounding.

- truncation - trunc() - prostě vezme hodnotu před . -> jedno jaká hodnota, 10.4-10,10.8-10 atd.., u negativních funguje stejně
- int funkce používá trun funkci pokud vytváří integer a zadáme mu float, tzn. můžu použít int místo trunc() funkce
- floor je trochu zamotanějš, bere nejmenší číslo po x -> zde to bude odlišné od záporných hodnot
- tzn. 10.4 - floor je 10 ale u -10.4 je floor 11
- takže u pozitivních čísel je floor a truncation shodné u negativních hodnot je to odlišné
- floor() funkce tez z knihovny math
- Ceiling je opak floor, bere tedy nejmensi tzn. 10.4 - 11, -10.4 - -10

### Coding part

In [146]:
from math import trunc, floor, ceil

In [142]:
trunc(10.3), trunc(10.5), trunc(-10.8)
# neni co resit 

(10, 10, -10)

In [145]:
floor(10.3), floor(10.5), floor(-10.8)
# zde je tedu rozdil u negativnich cisel

(10, 10, -11)

In [ ]:
floor(10.3), floor(10.5), floor(-10.8)

In [147]:
ceil(10.3), ceil(10.5), ceil(-10.8)
# a tady to mame obracene

(11, 11, -10)

In [149]:
int(10.8) #trunc netreba pouzivat

10

### Floats Rounding teorie

Round funkce round(x,n=0), n je v podstate na kolik chceme zaokrouhlit. round(x) -> prevedeme na int

- round 1.3 -> round(x) = 1 a integer, 0.23 k hodnote 1 a 0.77 k hodnote 2
- round 1.3 -> round(x,0) = 1.0 a vysledek float
- round(x, n=1) -> zaokrouhlujeme na 10 na -1 tedy na 0.1 = 1.23 na desetiny 0.3 k 1.2 a 0.7 k 1.3 takze vysledek je 1.2
- predtim bylo 10 na 0 -> 1 a tudiz zaokrouhleni na jednotky cele
- round(x, n=-1) tak menim 10 na --1 takze 10 na 1 - zaokrouhliji na desitky --> pokud by bylo x 18.3 zaokrouhli na 20
- x = 1,25 a chceme na desetinu pote mame 0.05 a 0.05 co pote? Zaokrouhluje pryc od nuly tzn na 1.3 ale u zaporu tez tzn -1.25 na -1.3
- nicmene mame zde bankers rounding takze zaokrouhluje na suda - pokud ma lichou hodnotu zaokrouhli opacne tzn 1.25 na 1.2 a 1.35 na 1.4, 15 na 20, 25 na 20 
- Bankovni rounding právě proto ze 5ka by se zaokrouhlila nahoru --> vetsi uroky apod..nejednotne
- 0.5 1.5 2.5 --> 1.5 je prumer, 2 prumer pri klasickem zaokrouhleni, 1,33 prumer -> presnejsi 'lesss bias'
- int(x +0.5) pridavani 0.5 aby dochazelo k zaokrouhlovani away from zero --> nefunguje u zapornych cisel ale mozne to je taktez
- pokud chceme opravdu zaokrouhlit away from zero musime nejdrive provest absolutni hodnotu pro x - a pote navratit znak --> vykryjeme tak zaporne hodnoty, pouziva pro to funkci math.copysign

### Coding part

In [150]:
#round funkci neni treba importovat je defaultne v pythonu

In [152]:
a = round(1.9)
a, type(a)
#tzn pokud nezadame nic prehodine na integer 

(2, int)

In [153]:
a = round(1.9,0)
a, type(a)
# jakmile zadam parametr pro n jiz prevede na floaut

(2.0, float)

n > 0

In [154]:
round(1.8888,3), round(1.8888,2),round(1.8888,0)
#jasna ukazka zaokrouhleni 

(1.889, 1.89, 2.0)

n < 0

In [156]:
round(888.88,1),round(888.88,0),round(888.88,-1),round(888.88,-2), round(888.88,-3), round(888.88,-4)
#na desetiny, jednotky, desitky, stovky, tisice, desetitice -- ma to blize k 0 nez k 10 000

(888.9, 889.0, 890.0, 900.0, 1000.0, 0.0)

Ties

In [158]:
round(1.25,1), round(1.35,1) #bankovni zaokrouhlovani - k sudym cislum

(1.2, 1.4)

In [159]:
round(-1.25,1), round(-1.35,1) #shodne i pro zaporne hodnoty

(-1.2, -1.4)

Decimals maji vice technik pro zaokrouhleni u floatu se to musi dotvorit.

In [161]:
#musim si vytvorit spesl metodu pokud chci away from zero
def _round(x):
    from math import copysign
    return int(x+0.5*copysign(1,x))
    #v podstate vezme x da do absolutni hodnoty a prida znamenko

In [162]:
round(2.5), _round(2.5)
#vidime zde 

(2, 3)

### Decimals teorie

Řeseni pokud je nasim cilem vyhnout se odhadum a mit 'presna' cisla. Neresi se pres Fraction protoze je to narocnejsi na pamet i vykon. Napriklad v bankovnictvi je nutne mit vse presne jinak by se pri x transakci ve vysledku nascitali velike rozdily.

- mame pravidla pro zaokrouhlovani
- dale muzeme specifikovat presnost pri aritmetickych operaci
- defaultni nastaveni je globalni context ale muzu si nastavovat svuj localni context tzn. pravidla pro presnost a zaokrouhlovani
- import decimal pro import, decimal.getcontext() nas defaultni
- decimal.localcontext(ctx=None) --> dostaneme context manager
- ctx.prec, ctx.rounding pro get a set - uvidime v kodu
- mame zde vice hodnot ktere muzeme definovat a zmenit tak nastaveni napr ROUND_UP apod.
- decimal.getcontext().rounding = decimal.ROUND_HALF_UP - napriklad takto
- LOCAL uz je trochu narocnejsi uvidime v kodu - jede na with logice ..tzn context manageru

### Coding Part

In [179]:
import decimal
from decimal import Decimal

In [180]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

Nas globalni context tzn. tak jak to máme nastavené.

In [181]:
decimal.getcontext().rounding
#zjistujeme hodnotu zaokouhleni

'ROUND_HALF_EVEN'

In [190]:
decimal.DefaultContext.prec = 6
#nastavuji hodnotu pro global context

In [191]:
q_ctx = decimal.getcontext()

In [192]:
type(q_ctx)

decimal.Context

In [188]:
q_ctx.rounding = decimal.ROUND_HALF_UP
decimal.DefaultContext.rounding = decimal.ROUND_HALF_UP

In [193]:
decimal.getcontext()

Context(prec=6, rounding=ROUND_HALF_UP, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

DefaultContext musim dodavat abych provedl zmeny

In [196]:
#pouze nastavim do zakladu
decimal.DefaultContext.rounding = decimal.ROUND_HALF_EVEN
decimal.DefaultContext.prec = 28

In [197]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [198]:
type(decimal.localcontext()) #takto pouze lokalni context
#vraci managera global dodava context = musime ladit

decimal.ContextManager

In [201]:
type(decimal.getcontext()) #zde vidime rozdil

decimal.Context

Tzn lokalni musime resit trochu jinak

In [202]:
with decimal.localcontext() as ctx:
    print(type(ctx))
    print(ctx)
# takhle se do contextu dostanu

<class 'decimal.Context'>
Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])


In [205]:
x = Decimal('1.25')
y = Decimal ('1.35')

In [206]:
with decimal.localcontext() as ctx:
    ctx.prec = 6
    ctx.rounding = decimal.ROUND_HALF_UP
    print(ctx)
    print(decimal.getcontext())
    print(round(x,1))
    print(round(y,1))
print(round(x,1))
print(round(y,1))

Context(prec=6, rounding=ROUND_HALF_UP, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])
Context(prec=6, rounding=ROUND_HALF_UP, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])
1.3
1.4
1.2
1.4


Takze takto to jsme schopni menit. Uvnitr with codu definujeme lokalni context. Funkce getcontext vraci context v kterym je, tzn. ve with bloku navraci lokalni context.

Dale vidim zaokrouhleni na urovni v contextu a global contextu. Tzn uvnitr zaokrouhlujeme nahoru a pote podle banking na even.

### Constructors and Context teorie

Nase classa decimal tzn Decimal(x) muzeme zapsat vice zpusoby jako integer -- Decimal(10), string '0.1', tuples ale to je trochu zamotane. Floats radeji nepouzivat v podstate to zaeviduje s jejich nepresnosti.

- TUPLE CONSTRUCTOR
- 1.23 zapiseme jako +123 x 10 na -2
- -1.23  jako -123 x 10 na -2
- znamenko - cislice - exponent
- takze v tuplu se decimal zapisuje jako (sign(c1,c2...),exponent)
- stim ze s je 0 pokud mame kladne cislo
- s davame jako 1 pokud mame zaporne cislo
- -3.1415 zapiseme v tuple jako (1,(3,1,4,1,5),-4)
- exponent abychom posunuli carku v podstate o 4

Presnost constructu ovlivnuje matematicke operace ale construktor samotny neovlivnuje. 

- a = Decimal('0.12345'), b = Decimal('0.12345')
- DO KONTEXTU OBOJI ULOZI TAKTO S PRESNOSTI 5
- c = a+b => 0.25 ALE PRI VYPOCTU PRESNOST RESI

### Coding part

In [207]:
import decimal
from decimal import Decimal

In [210]:
Decimal(10), Decimal('10.1') #integer a string zapis

(Decimal('10'), Decimal('10.1'))

In [220]:
t = (0, (3,1,4,1,5), -4) #nejdriv musim takto
Decimal(t)
Decimal((0, (3,1,4,1,5), -4))
# pokud chci zapsat musim dalsi zavorky jinak netvorim tuple

Decimal('3.1415')

In [221]:
Decimal((1, (3,1,4,1,5), -4))#1ka pro zapornou hodnotu

Decimal('-3.1415')

In [222]:
format(0.1, '.25f')

'0.1000000000000000055511151'

In [223]:
Decimal(0.1) #nepouzivat float jako parametr pro Decimal

Decimal('0.1000000000000000055511151231257827021181583404541015625')

In [224]:
Decimal(0.1) == Decimal('0.1') #strig je ale funkcni

False

In [225]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [228]:
decimal.DefaultContext.prec = 6
# nastavim presnost kontextu 

In [229]:
a = Decimal('0.123456789')

In [230]:
a #ale i pres prenost 6 to vidi spravne

Decimal('0.123456789')

In [232]:
decimal.DefaultContext.prec = 2
a = Decimal('0.12345')
b = Decimal('0.12345')
c = a+b
print(a,b,c)
#Ale zde vidime ze pro vypocty se presnost meni
# tzn secte a zaokrouhli 

0.12345 0.12345 0.25


In [235]:
decimal.DefaultContext.prec = 6
a = Decimal('0.12345')
b = Decimal('0.12345')
print(a+b)
with decimal.localcontext() as ctx:
    ctx.prec =2
    c = a + b
    print(f'{c} with local context')
print(f'{c} with global context')

0.24690
0.25 with local context
0.25 with global context


Nejdrive jsme pocitali s presnosti 6 a tudiz mame 0.24690 ale pote doslo k prechodu na presnost 2. Pokud to ale znovu neprepisi stale si pamatuje hodnotz z local contextu.

### Decimal Math operations teorie

Nektere operace se chovaji odlisne od floatu. Zejmena div a mod tzn // a %.

 - a // b = floor(a/b) pro integers ale pro decimal trunc(a/b)
 - 10// 3 = 3, pro decimal vysledek 3
 - 10 // 3 = -4 ale pro decimal vysledek -3 
 - tzn. to mame trochu mindfuck ale dejme tomu
 - 10 // 3 => 10 - 3 = 7, 7 - 3 = 4, 4 -3 = 1 STOP vysledek je mensi nez delitel a odecitali jsme 3x proto 3
 - -10 // 3 => mame minus odebereme negativni znamenko a potom jej vratime --> 3x a jsme na -3
 - je to stejna operace jako trunc metoda
 - Nase rovnice je stale stejna a stale se rovna
 - -135 // 4 pro integer -34 a pro decimal -33 takze zbytek je 1 a -3 => rozdilny zbytek ale rovnice shodna
 - math modul muzeme pouzivat na decimals ale zmeni si decimals na float a tim ztraci smysl je vubec pouzivat - lepsi se math modulu vyhnout

### Coding part

In [236]:
import decimal
from decimal import Decimal

// and %

n = d*(n//d) + (n%d)

In [238]:
x, y = 10, 3 
print(x//y, x%y)
# delime inetger mame 3x a zbytek mame 1
# deleni inetegeru je 3 protoze nema logicky ..

3 1


In [239]:
x, y = Decimal(10), Decimal(3)
print(x//y, x%y)
#s decimal mame stejne hodnoty - mame +

3 1


In [243]:
x, y = -10, 3 
print(x//y, x%y)
# 3*-4 je -12 + 2 je -10

-4 2


In [244]:
x, y = Decimal(-10), Decimal(3)
print(x//y, x%y)
# 3* -3 a  -1 je tez -10

-3 -1


V podstate jen ukazka nasi teorie jen na prikladech. Spise potrebuji vedet neco ve stylu - aha bacha na to. Ale umet to pravdepodobne nemusim momentalne umet :)

In [253]:
decimal.DefaultContext.prec = 28
a = Decimal('0.1')
print(a.ln()), print(a.exp()), print(a.sqrt())
#tohle jsou nase operace v ramci decimal classy

-2.302585092994045684017991455
1.105170918075647624811707826
0.3162277660168379331998893544


(None, None, None)

In [249]:
import math

In [255]:
math.sqrt(a)
#0.3162277660168379331998893544
#0.31622776601683794
#jina presnost takze mame mensi rozdily protoze math modul prehodi moji hodnotu na floats

0.31622776601683794

In [256]:
x = 2
x_dec = Decimal(2)

In [259]:
root_float = math.sqrt(x)
root_mixed = math.sqrt(x_dec)
root_dec = x_dec.sqrt()

In [264]:
print(format(root_float, '1.27f'))
print(format(root_mixed, '1.27f'))
print(root_dec)
## zajimave odlisnost math od opaku

1.414213562373095145474621859
1.414213562373095145474621859
1.414213562373095048801688724


In [265]:
print(format(root_float*root_float, '1.27f'))
print(format(root_mixed*root_mixed, '1.27f'))
print(root_dec*root_dec)
#prvni a druhy ma shodne vysledkz protoze ikdyz mame decimal a float
# math knihovna pouzije decimal jako float takze to mame shodne
#ale co je pekne nase decimal operace je mnohem zdatnejsi ==> vetsi presnost

2.000000000000000444089209850
2.000000000000000444089209850
1.999999999999999999999999999


In [268]:
x_dec_not_floated = Decimal('0.01')
x_dec_floated = Decimal(0.01)
print(x_dec_not_floated)
print(x_dec_floated)
#Tady to vidime tak trochu presne jak nam to zfusuje na float

0.01
0.01000000000000000020816681711721685132943093776702880859375


### Decimals Performance Considerations

- obecne je to pracnejsi, vytvareni Decimal neni tak snadne jako float
- chybi nektere funkce co jsou v math modulu zejmena triginometricke funkce
- je zde vetsi memory overhead
- jsou pomalejsi 'relativne' oproti floats
- pokud presnost neni dulezita neni treba pouzit decimals

In [270]:
from decimal import Decimal
import sys

In [271]:
a = 3.1415
b = Decimal('3.1415')

In [272]:
sys.getsizeof(a)

24

In [274]:
sys.getsizeof(b)
#vice nez 4x vetsi pamatova narocnost

104

In [276]:
import time

Dame si mensi benchmark

In [290]:
def run_float(n=1):
    for i in range(n):
        a = 3.1415
        
def run_decimal(n=1):
    for i in range(n):
        a = Decimal('3.1415')
#pouze benchamrk skippnu nechce mi to pocitat ma to problemy se smickou 

Vysledek je jasne. Decimals pouzivaji vice pameti a vice vykonu a to opravdu o kotel. Prenost si zada svou obet.

### Complex Numbers teorie a coding

Tuhle část můžu opravdu přeskočit. Až moc specifické.

### Boolenas teorie

Bool classa je subcalssa int classy a přidává pár dalších navíc jako and, or apod.

In [291]:
issubclass(bool,int) #test shodz

True

In [294]:
isinstance(True, bool), isinstance(False, bool), isinstance(True, int)

(True, True, True)

True a False je v podsta jen nalepka pro integer 1 a 0 s extra funkcionalitou.

Jelikoy true a false ma jen 1 a 0 maji porad stejnou hodnotu. Tzn. is a == jsou v tomto pripade shodne protoze hodnoty ani reference se nemeni.

In [295]:
int(True), int(False)

(1, 0)

True a 1 a False a 0 NEJSOU stejne objek. Tzn maji jine ID adress. 

In [299]:
True == 1, True is 1, False is 0, False == 0

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-299-be3389b51640>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  True == 1, True is 1, False is 0, False == 0
<ipython-input-299-be3389b51640>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  True == 1, True is 1, False is 0, False == 0


(True, False, False, True)

In [300]:
True > False # trocha zabavz, 

True

In [303]:
(1 == 2) == False, (1 == 2) == 0
# prvni cats vyhodnoti jako False a False rovna se False takze True

(True, True)

In [304]:
True + True + True #opravdu hodne zabavne 

3

In [305]:
-True

-1

- Boolena constructor, bool(x), True pokud x je true a naopak
- Kazda classa v pythonu ma Bool hodnotu - takze takto mohu volat True a False hodnotu pro nase objekty
- Samozrejme to ma smyl zejmena pokud mam custom classy a chci je takto labelovat
- Integer ma napriklad False pokud je 0 a jinak ma True

In [307]:
bool(0), bool(24) #jen 0 je false

(False, True)

In [309]:
id(False),id(True)

(140709391161200, 140709391161168)

In [310]:
type(3<4)

bool

In [312]:
id(3<4) #id je stejne jako id True protoze je to True 
#tzn False a True hodnoty maj jedny jasne adresy

140709391161168

In [315]:
bool(0), bool(1), bool(5),bool('lkjlk')
#false mame zatim jen u  uvidime dale

(False, True, True, True)

### Booleans True values teorie

- vsechny objekty jsou vyhodnoceny jako True az na par vyjimek
- None
- False
- 0
- prazdne sekvence jako listy apod
- prazdne dictionary apod
- custom classy dle nasi implementace

Pokud zavolame bool(x) tak python vola x.__bool__() dunder metodu a pokud ji nenajde tak __len__ metodu a pokud i tak nic nenajde tak navrati True. 

In [316]:
def __bool__(self):
    return self != 0
#nejak takto by mohla fungovat

In [319]:
bool([]), bool(None), bool(0), bool({})
#zde to u listu a dt bere v potaz prave len

(False, False, False, False)

In [320]:
from decimal import Decimal
from fractions import Fraction

In [321]:
bool(Fraction(0,1)) #0 lomeno 1 je 0

False

In [322]:
bool(Decimal('10.5'))

True

In [323]:
bool('String je taktez sekvence a proto se resi pres len metodu')

True

In [324]:
a = [1,2,3]
if a is not None and len(a) > 0:
    print(a[0])
else:
    print('Nothing to see here')

1


In [327]:
a = None # nebo a = 0, ci a = []
if bool(a):
    print(a[0])
else:
    print('Nothing to see here')

Nothing to see here


V podstate cela logika True a False je vyse zminena. Resi jestli to neni 0 nebo none a nebo jestli ma nejakou delku :).

### Booleans Operator, precedence and short circuit evaluation

Operátory jsou jasné NOT AND OR. Často se používá graf u obvodů - otevřený obvod FALSE, zavřený tzn průchozí TRUE. OR se zobrazuje jako dva switche paralerně. AND jsou dva switche za sebou a pokud nemají oba open neboli TRUE nic nejde skrz. 

- Precedence podle vlivu co nejdrive a co na zaver
- ()
- <> <= >= != in ins
- not
- and
- or
- Tohle je pomerne dobre si pamatovat :), na poradi zalezi, obecne spravne udavat zavorky je hlavni, kod bude vic readable

In [328]:
#Short circuting ukazka:
#if symbol in watch_list:
#    if price(symbol) > threshold:
#        do something
#==> jede na dve faze narocnejsi kod
#
#V ramci short circuting - tzn utnuti na zacatku co to jde
#
#if symbol in watch_list and price(symbol) > threshold:
#    do something
#
#==> druhy priklad je castecne zdatnejsi neprojede vse ale cast ukonci na zacatku a usetri tak cas

In [329]:
# chceme provest akci kontrolovat zda prvni znak je digit
# if name[0] in string.digits:
#     do something
# tento kod se rozbije pokud je jmeno None nebo prazdny string

# spravne to je
# if name and name[0] in string.digits:
#           do something
#NYNI to overuje zda je name true neboli zda neni none nebo ma len a pote prejde na druhou cast
# a zacina to luxusne zapadat :)

### Coding part

not --> and --> or

In [330]:
True or True and False

True

In [331]:
True or (True and False)
#takto to prehledneji..and se zde vzhodnotilo samo

True

Short circuting

Jakmile ukonci obvod nemusi vyhodnocovat dalsi casti a setrime cas

In [332]:
a = 10
b = 2

if a/b > 2:
    print('a is at least twice b')

a is at least twice b


In [335]:
b = 0
if b and a/b >2:
    print('do something')
    
#prvni priklad nad hodi chybu pokud delime 0
#takto tomu ale predejdeme
#nemam zadne podchycovani zerodivision apod..
#dokonce poresim i problem s None
#moc pekne :)

In [336]:
import string

In [337]:
string.ascii_uppercase

'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

In [338]:
string.ascii_letters

'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'

In [339]:
string.digits

'0123456789'

Pouze ukazka sikovne knihovnicky se znaky.

In [340]:
name = ''
if name[0] in string.digits:
    print('Name cannon start with a digit')

IndexError: string index out of range

Neumime pokyt prazdny string ani none spravne takto

In [341]:
name = ''
if name and name[0] in string.digits:
    print('Name cannon start with a digit')

Asi nejkrasnejsi dva radky za celou lekci. Konecne to dava perfektni smysl :)

### Boolean Operator teorie 2

Jelikoy kazdy objekt disponuje bool hodnotou muzeme vse porovnavat a nemusime ani formou bool(x) and bool(y) staci zapisovat formou X and Y, ale otazka je co zde ziskavame?

Tohle je trochu zamotanejsi takze:

- OR
- if X je Tru navratime X nebo Y, povazujme hodnoty dale zapsane jako X,Y
- 0 0 X je false vracime Y tedy 0
- 0 1 X je false vracime Y tedy 0
- 1 0 a 1 1 X je pravda vracime 1
- jenze Y nemusime ani cist dokud X neni pravda
- LEPSI JE TEDY:
- POKUD JE X Pravda vrat X jinak vrat Y

Nyni se presuneme na A

- AND
- IF X IS FALSZ returns X otherwise retunrs Y
- 0 0 X je nepravda vracime X a tedy 0
- 0 1 X je nepravda vracime X a tedy 0
- 1 0 X je pravda vracime Y a tedy 0
- 1 1 Y je pravda vracime Y a tedz 1
- POKUD X JE NEPRAVDA VRAT X JINAK VRAT Y

V podstate AND a OR logika je ted trochu jina. Zatimco pokud neco porovnavame mame trochu jina pravidla nez pokud porovnavame dva objekty ktere samy o sobe maji TRUE/FALSE hodnotu.

In [343]:
s1,s2,s3 = '','',''
a = s1 or s2 or s3 or 'N/A'
#Mame OR operatory a short circuit neboli 
#POKUD JE NECO PRAVDA VRATI TO JINAK JDE DAL
#V TOMTO PRIPADE U OPERATORU Y LEVA DO PRAVA VZHODNOCUJEME
#TAKZE COKOLIV CO BUDE PRAVDA JAKO PRVNI SE NAVRATI

Takze tohle je pekne. Chceme obdrzet prvni znak ze jmena ci nic pokud chybi.

In [346]:
#if s:
#    return s[0]
#else:
#    return ''
#noobie zapis

#return s and s[0]
#profi zapis jelikoz mame and tak vrati s jen pokud bude falsy a tudiz none nebo bez delky
#jinak navrati prvni znak
# return (s and s[0]) or ''
# tohle uz je tezka stylovka dokonce si urcime hodnotu kterou nam ma navratit pokud nastane none pripad

NOT OPERATOR

- not x 
- True if x is falsy
- False if x is truthy
- [] falsy -> not [] -> True ..jasne netreba resit

### Coding Part

X or Y: if X is truthy, returns X, otherwise evaluates Y and returns it

In [347]:
'a' or [1,2]#prvni je truthy navrati jej

'a'

In [348]:
'' or [1,2] #nepravdive navrati to druhe

[1, 2]

In [349]:
'' or '' or '' or 'a' # hehe

'a'

In [350]:
1 or 1/0 #nemam zerodivison error vyhodnoti prvne 1 a ukonci tzn. short circuting

1

In [351]:
0 or 1/0  #nenastane pravda short circuting nenastane

ZeroDivisionError: division by zero

X and Y: if X is falsy, return X, otherwise evaluate Y and return it

In [353]:
print(None and 100)

None


In [354]:
[] and [0]

[]

Mame zde mensi opak u and, v podstate obdrzujeme prvni false hodnotu

OR POUZIJI POKUD CHCI VYDOLOVAT TRUTH HODNOTU A IGNOROVAT FALSE HODNOTY pokud ale nejsou truth ziskat FALSE

AND POUZIJU POKUD CHCI VZDOLOVAT FALSE HODNOTU A IGNOROVAT TRUTH HODNOTY pokud ale nejsou FALSE ziskat TRUTH

In [355]:
a,b = 2,4
if b == 0:
    print(0)
else:
    print(a/b)

0.5


A ted trochu styloveji.

In [356]:
a = 2 
b = 0
print(b and a/b)

0


Aneb zero divison error handling..

In [357]:
s1, s2, s3 = None, '', 'abc'

In [361]:
((s1 and s1[0]) or 'ok '), ((s2 and s2[0]) or ' '), ((s3 and s3[0]) or ' ')

('ok ', ' ', 'a')

Co asi dela tento krasny kod :]. 
- s1 and s1[0] - Pokud je prvni falsy navrati ji jinak navrati druhou
- takze navrati None, None a pote abc
- a nas or navrati prvni True hodnotu 
- tzn navrati abc protoze uz je true ale pro ostatni dve navrati bud true hodnotu nebo to co zbyde
- proto v prvnim mam true hodnotu a v druhem false ale oboji se navrati

In [364]:
not True, not False, not bool('abc'), not ('abc')
#v podstate ~ logic :]

(False, True, False, False)

### Comparison Operators

- binarni operatory napriklad + potrebuje A a B
- identifikacni operatorz - is a is not
- value comparisons == !=
- ordering comparison > < >= >=
- memebership operations - not a not in

Muzu mezi sebou porovnavat napriklad float s decimal, integrem apod..Krome komplexnich cisel na ty je treba dat si bacha.

a == b == c je stejne jako a == b and b == c atd.

Trochu to zamotame

1 < 10 > 4 < 5 --> 1 < 10 and 10 > 4 and 4 < 5..v podstate to jde poporade takze True 

In [365]:
0.1 is (3+4j) #jen ukazka ze nemam radi komplexni cisla

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-365-aadb6322e9fe>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  0.1 is (3+4j)


False

In [367]:
import string

In [368]:
'A' < 'a' < 'z' > 'Z' in string.ascii_letters
#sekvence charakteru - uupercase je dale nez lovercase
#takze tak trochu pro zabavu - porovnavame i stringy

True

Tak a to je vse k useku 4 o datovych typech. Vsechno velice prinosne ale nastesti zatim bez vetsich problemu v chapani.